internal/libs/mathslib/calc.go

```diff
@@ -23,9 +28,43 @@ func Deviates(x1, x2 *big.Int, ppb int64) bool {
 	return diff.Cmp(big.NewInt(ppb)) > 0 // diff > ppb
 }
 
-// CalculateUsdPerUnitGas returns: (sourceGasPrice * usdPerFeeCoin) / 1e18
-func CalculateUsdPerUnitGas(sourceGasPrice *big.Int, usdPerFeeCoin *big.Int) *big.Int {
-	// (wei / gas) * (usd / eth) * (1 eth / 1e18 wei)  = usd/gas
-	tmp := new(big.Int).Mul(sourceGasPrice, usdPerFeeCoin)
-	return tmp.Div(tmp, big.NewInt(1e18))
+// CalculateUsdPerUnitGas accepts source gas price denoted in chain-specific price units,
+// and usdPerFeeCoin denoted in 1e18 USD per 1e18 smallest token unit (the FeeQuoter notation),
+// and returns the source gas price denoted in 1e18 USD per gas unit.
+func CalculateUsdPerUnitGas(
+	sourceChainSelector ccipocr3.ChainSelector,
+	sourceGasPrice *big.Int,
+	usdPerFeeCoin *big.Int) (*big.Int, error) {
+
+	family, err := chainsel.GetSelectorFamily(uint64(sourceChainSelector))
+	if err != nil {
+		return nil, fmt.Errorf("failed to get chain family for selector %d: %w", sourceChainSelector, err)
+	}
+
+	switch family {
+	case chainsel.FamilyEVM:
+		// In EVM, sourceGasPrice is denoted in wei/gas.
+		// Eth has 18 decimals, usdPerFeeCoin represents 1e18 USD/1e18 wei.
+		// To get 1e18 USD/gas, we multiply sourceGasPrice by usdPerFeeCoin and divide by 1e18.
+		tmp := new(big.Int).Mul(sourceGasPrice, usdPerFeeCoin)
+		return tmp.Div(tmp, big.NewInt(1e18)), nil
+
+	case chainsel.FamilySolana:
+		// In SVM, sourceGasPrice is denoted in microlamports/cu, or 1e-15 SOL/cu.
+		// We need to multiply by 1e3 to be in units of 1e-18 SOL/cu, equivalent to wei/gas.
+
+		// Sol has 9 decimals, usdPerFeeCoin represents 1e18 USD per 1e9 whole sol, or 1e18 USD/1e27wei.
+		// We need to divide by 1e9 to be in units of 1e18 USD/1e18 wei, equivalent to EVM model.
+
+		// To keep precision, we want to divide at the very end.
+		// The above cancels out, the net result is we need to further divide by 1e6,
+		// or 1e18 * 1e6 = 1e24 in total.
```

Comment:
I feel like the formula that was there before was pretty clear and mirrors the code almost exactly:

```
// (micro lamport / compute units) * (usd18 / 1e9 sol) * (1e9 sol / 1e24 micro lamport)  = usd18/cu
// or
// (microlamport / cu) * (usd18 / 1e18 lamport) * (1 lamport / 1e6 microlamport)
```